File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.math.BigInteger;
import java.util.FormatFlagsConversionMismatchException;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Stack<Character> stack;
        StringBuilder sb = new StringBuilder();
        while(true) {
        	stack = new Stack<>();
        	boolean isBalanced = true;
        	boolean isEnd = false;
        	String str = br.readLine();
        	if(str.equals(".")) break;
        	char[] arr = str.toCharArray();
        	for(char c : arr) {
        		if(c == '[') {
        			stack.push(c);
        		} else if(c == ']') {
        			if(!stack.isEmpty() && stack.peek() == '[') stack.pop();
        			else {
        				isBalanced = false;
        				isEnd = true;
        				break;
        			}
        		} else if(c == '(') {
        			stack.push(c);
        		} else if(c == ')') {
        			if(!stack.isEmpty() && stack.peek() == '(') stack.pop();
        			else {
        				isBalanced = false;
        				isEnd = true;
        				break;
        			}
        		}
        	}
        	
        	if(isEnd) {
        		sb.append(isBalanced?"yes\n":"no\n");	
        		continue;
        	}
        	
        	isBalanced = stack.size() > 0 ? false : true;
        	sb.append(isBalanced?"yes\n":"no\n");
        }
        
        

        System.out.println(sb);
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

if __name__ == "__main__":
    while True:
        s = input()
        ans = 'yes'
        if s == ".":
            break
        
        stack = []
        
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '[':
                stack.append(s[i])
            elif s[i] == ')':
                if len(stack) == 0 or stack[-1] != '(':
                    ans = 'no'
                    break
                stack.pop()
            elif s[i] == ']':
                if len(stack) == 0 or stack[-1] != '[':
                    ans = 'no'
                    break
                stack.pop()
        if len(stack) != 0:
            ans = 'no'
            
        print(ans)